### PR TITLE
Renesas ra2 

### PR DESCRIPTION
**Pull request by avolkov-1221**
This is an initial support of Renesas RA2L1 chips family and EK-RA2L1 eval board. Unlike the Renesas port, it does not require additional HAL modules.

### CODE REVIEW

File: soc/renesas/ra/ra2/Kconfig.defconfig
Comment (gmarull) [inline soc/renesas/ra/ra2/Kconfig.defconfig:16]: -1, drivers need to select

**Review by gmarull — CHANGES_REQUESTED** [review]

File: soc/renesas/ra/ra2/Kconfig.defconfig
Comment (avolkov-1221) [reply soc/renesas/ra/ra2/Kconfig.defconfig:16]: Could you explain? PINCTRL is a subsystem, not a driver.

File: include/zephyr/drivers/pinctrl/pinctrl_ra2.h
Comment (avolkov-1221) [reply include/zephyr/drivers/pinctrl/pinctrl_ra2.h:21]: pinctrl_ra_get_pin is required for the GPIO driver. Other -  helpers for our software, can be removed.

**Review by soburi — CHANGES_REQUESTED** [review]
There are several PRs to add support for RA2, including my own.
In my opinion, having multiple implementations makes maintenance difficult, so I think we should consolidate them.
I would also like to align RA4M1, which I created previously, with this one.

It would be easier to discuss if you submitted it as a patch to the existing drivers rather than adding an alternative driver.

File: drivers/clock_control/clock_control_ra2_hoco_osc.c
Comment (avolkov-1221) [reply drivers/clock_control/clock_control_ra2_hoco_osc.c:26]: Yes sure, but not HOCO one. Could you please check the logic of "virtual" sysclock in "clock_control_ra2_sysclk.c" and "clock_control_ra2_root_osc.c" ? Also could you check the ra2l1.dts ?

**Comment by avolkov-1221** [comment]
> There are several PRs to add support for RA2, including my own. 

Sorry, but could you give me this PR number ? I only see my and Renesas PRs.

>In my opinion, having multiple implementations makes maintenance difficult, so I think we should consolidate them. I would also like to align RA4M1, which I created previously, with this one.
> 
> It would be easier to discuss if you submitted it as a patch to the existing drivers rather than adding an alternative driver.It will definitely be good to at least combine our work. I didn't do this because I don't have the RA4M1 hardware and can't test it.

It will definitely be good to at least merge our works. I didn't do this because I don't have the RA4M1 hardware and can't test it. Part of our implementations are close enough and didn't depends on external modules, like Renesas port do.

File: drivers/clock_control/clock_control_ra2_hoco_osc.c
Comment (avolkov-1221) [reply drivers/clock_control/clock_control_ra2_hoco_osc.c:26]: Oh, sorry, I didn't quite understand what you were talking about at first. At first I tried to use the subsystem, but, unfortunately, due to the Renesas hardware implementation and Device Tree syntax, this turned out to be extremely inconvenient: it was not clear how to define the "clock" property in the peripheral nodes. In the case of "subsystem" there must be at least 2 fields in the reference, but some clocks such as MOSC cannot be used this way.

File: include/zephyr/drivers/pinctrl/pinctrl_ra2.h
Comment (avolkov-1221) [reply include/zephyr/drivers/pinctrl/pinctrl_ra2.h]: Fixed

**Comment by avolkov-1221** [comment]
> There are several PRs to add support for RA2, including my own. In my opinion, having multiple implementations makes maintenance difficult, so I think we should consolidate them. I would also like to align RA4M1, which I created previously, with this one.
> 
Hello @soburi ,

This PR now is ready to merge with your code (I mean RA4 + the latest ICU), so I'm also ready to discuss any suggestions/objections.

**Comment by KhiemNguyenT** [comment]
Hi @avolkov-1221 , @soburi , as the comment in top post, this port does not use HAL drivers in hal_renesas, which will cause conflict with the new implementation that we are trying to achieve on Renesas RA8 port. In short, we intend to leverage HAL in RA FSP, for code reusability. Of course, some exception may be considered separately.  The implementation should be streamlined across Renesas MCU port, not just RA8 (cc @soburi , we may discuss more on Discord if you like).

**Comment by avolkov-1221** [comment]
> Hi @avolkov-1221 , @soburi , as the comment in top post, this port does not use HAL drivers in hal_renesas, which will cause conflict with the new implementation that we are trying to achieve on Renesas RA8 port. In short, we intend to leverage HAL in RA FSP, for code reusability. Of course, some exception may be considered separately. The implementation should be streamlined across Renesas MCU port, not just RA8 (cc @soburi , we may discuss more on Discord if you like).

I'm in Discord too. Check out my message in the renesas channel.

**Comment by avolkov-1221** [comment]
@ydamigos
Yannis, I don't understand whats happened? Why do you reassigned this PR to the unrelated person?

**Comment by avolkov-1221** [comment]
> Hi @avolkov-1221 , @soburi , as the comment in top post, this port does not use HAL drivers in hal_renesas, which will cause conflict with the new implementation that we are trying to achieve on Renesas RA8 port. In short, we intend to leverage HAL in RA FSP, for code reusability. Of course, some exception may be considered separately. The implementation should be streamlined across Renesas MCU port, not just RA8 (cc @soburi , we may discuss more on Discord if you like).

@KhiemNguyenT quality of FSP code din't allow to use it.

**Comment by KhiemNguyenT** [comment]
> > Hi @avolkov-1221 , @soburi , as the comment in top post, this port does not use HAL drivers in hal_renesas, which will cause conflict with the new implementation that we are trying to achieve on Renesas RA8 port. In short, we intend to leverage HAL in RA FSP, for code reusability. Of course, some exception may be considered separately. The implementation should be streamlined across Renesas MCU port, not just RA8 (cc @soburi , we may discuss more on Discord if you like).
> 
> @KhiemNguyenT quality of the FSP code doesn't allow to use it.

@avolkov-1221 the discussion is on-going in Discord topic, I believe.
While we have 2 approaches to enable the support for Renesas RA devices, I think there are pros and cons of what should be the main path. As far as I observe, we did not re-invent a different approach, other vendors did similar things.
Now, come to some concerns about FSP code which we try to re-use in HAL layer. If there's no major blocking points for Zephyr Project as an open-source project, it's acceptable in Zephyr community, I think.

**Comment by ydamigos** [comment]
> @ydamigos Yannis, I don't understand whats happened? Why do you reassigned this PR to the unrelated person?

@KhiemNguyenT is one of the maintainers for Renesas RA Platforms (please check MAINTAINERS.yml), so I assigned him this PR.

**Comment by avolkov-1221** [comment]
> > @ydamigos Yannis, I don't understand whats happened? Why do you reassigned this PR to the unrelated person?
> 
> @KhiemNguyenT is one of the maintainers for Renesas RA Platforms (please check MAINTAINERS.yml), so I assigned him this PR.

Are you kidding me? It was you, who pushed him through as a maintainer, violating Zephyr's rules:
* He became a maintainer without a single verified commit, just based on trust in you.
* And the first PR from this team contained serious violations of many of Zephyr's code guide requirements, which shows that they lack experience. He shouldn't have become a "maintainer", nor should they have become "collaborators" without proper preparation.
* Then, as a "collaborator", not "maintainer", in an external project, you overstepped your authority and appointed him responsible for a competing PR.

Yannis, this looks like some dirty political game on your part to pushing vendor lock at any cost, rather than proper code collaboration within the community.

File: boards/renesas/ek_ra2l1/Kconfig.ek_ra2l1
Comment (nordicjm) [inline boards/renesas/ek_ra2l1/Kconfig.ek_ra2l1:5]:
```suggestion
```

File: boards/renesas/ek_ra2l1/ek_ra2l1_defconfig
Comment (nordicjm) [inline boards/renesas/ek_ra2l1/ek_ra2l1_defconfig:20]: Unset items can go

File: drivers/clock_control/clock_control_ra2_hoco_osc.c
Comment (nordicjm) [inline drivers/clock_control/clock_control_ra2_hoco_osc.c:25]: we don't put static on a different line in zephyr, goes on the same line

File: drivers/interrupt_controller/Kconfig.renesas_ra2
Comment (nordicjm) [inline drivers/interrupt_controller/Kconfig.renesas_ra2:8]: this should have a depends on to a soc family/vendor or similar

File: drivers/interrupt_controller/Kconfig.renesas_ra2
Comment (nordicjm) [inline drivers/interrupt_controller/Kconfig.renesas_ra2:25]: newline space after if and before endif

File: drivers/lpm/CMakeLists.txt
Comment (nordicjm) [inline drivers/lpm/CMakeLists.txt:1]: missing copyright/apache header

File: drivers/lpm/Kconfig
Comment (nordicjm) [inline drivers/lpm/Kconfig:9]: Since this is a new driver class, this needs to go through architecture review

File: soc/renesas/ra/ra2/Kconfig
Comment (nordicjm) [inline soc/renesas/ra/ra2/Kconfig:5]:
```suggestion
```

File: soc/renesas/ra/ra2/Kconfig
Comment (nordicjm) [inline soc/renesas/ra/ra2/Kconfig:15]:
```suggestion
```

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: drivers/clock_control/clock_control_ra2_hoco_osc.c
Comment (nordicjm) [reply drivers/clock_control/clock_control_ra2_hoco_osc.c:25]: Actually please just clang-format all of these files using the style included with zephyr

File: drivers/lpm/Kconfig
Comment (kartben) [reply drivers/lpm/Kconfig:9]: It doesn't even seem like there is a generic API (only lpm_ra2.h) so there is even more work needed there, I guess.

**Comment by KhiemNguyenT** [comment]
> > > @ydamigos Yannis, I don't understand whats happened? Why do you reassigned this PR to the unrelated person?
> > 
> > 
> > @KhiemNguyenT is one of the maintainers for Renesas RA Platforms (please check MAINTAINERS.yml), so I assigned him this PR.
> 
> @ydamigos Are you kidding me? It was you, who pushed him through as a maintainer, violating Zephyr's rules:
> 
> * He became a maintainer without a single verified commit, just based on trust in you.

Yannis helped me to accelerate the progress. And I got the approval from the Zephyr project admin. 
Being the maintainer, I help to 'maintain the code'. I may not have much exposures as other developers through.

> * And the first PR from this team contained serious violations of many of Zephyr's code guide requirements, which shows that they lack experience. He shouldn't have become a "maintainer", nor should they have become "collaborators" without proper preparation.

The PR are reviewed and approved by some community members, not a single member can make the decision.

> Yannis, this looks like some dirty political game on your part to pushing vendor lock at any cost, rather than proper code collaboration within the community.

This is a big statement. There are other PR from another contributor are accepted recently. This is one exception then you said "vendor lock", is it fair ?

File: drivers/lpm/Kconfig
Comment (avolkov-1221) [reply drivers/lpm/Kconfig:9]: > It doesn't even seem like there is a generic API (only lpm_ra2.h) so there is even more work needed there, I guess.

It's not a generic API certainly, but common part for all Renesas RA MCUs. And this IP is critical for almost all peripherals drivers, it's control the power. Current RA4 and RA8 implementations are working only because some peripherals are active by default after POR or MCU's specific LPM activation is hardcoded into the initialization.

File: drivers/clock_control/clock_control_ra2_hoco_osc.c
Comment (avolkov-1221) [reply drivers/clock_control/clock_control_ra2_hoco_osc.c:25]: > Actually please just clang-format all of these files using the style included with zephyr

No so easy. Unfortunately clang-format adore to break complex defs.

Comment (avolkov-1221) [reply drivers/clock_control/clock_control_ra2_hoco_osc.c:25]: > we don't put static on a different line in zephyr, goes on the same line

I can fix certainly the kernel's style "static" .

However, wouldn't be better to update following sentence in Zephyr's docos before?

From here https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style:
>
>...
>In general, follow the [Linux kernel coding style](https://kernel.org/doc/html/latest/process/coding-style.html), with the following exceptions:
>...
> When there are differences between the [Coding Style](https://docs.zephyrproject.org/latest/contribute/guidelines.html#id2) guidelines and the formatting generated by code formatting tools, the [Coding Style](https://docs.zephyrproject.org/latest/contribute/guidelines.html#id2) guidelines take precedence. If there is ambiguity between formatting tools and the guidelines, maintainers may decide which style should be adopted.

Because the LCG doesn't prohibit placing `static` on different line, and I didn't find such rule in the Zephyr's coding style guideline too.

File: drivers/lpm/Kconfig
Comment (nordicjm) [reply drivers/lpm/Kconfig:9]: then the drivers should be using the power management system

Comment (avolkov-1221) [reply drivers/lpm/Kconfig:9]: > then the drivers should be using the power management system

Unfortunately power management system can be disabled. In this case hardware will be disabled too forever (most of pm_* become dummy). It might seem odd to the user.

Comment (nordicjm) [reply drivers/lpm/Kconfig:9]:
```
config SOC_SERIES_blah
        select PM
```

File: soc/renesas/ra/ra2/Kconfig
Comment (KhiemNguyenT) [inline soc/renesas/ra/ra2/Kconfig:4]: It should be SOC_SERIES_RA2L1. See other PR for the naming convention.

File: soc/renesas/ra/ra2/Kconfig.defconfig
Comment (KhiemNguyenT) [inline soc/renesas/ra/ra2/Kconfig.defconfig:1]: Folder name should be ra2l1, rather than ra2.

File: soc/renesas/ra/ra2/Kconfig.soc
Comment (KhiemNguyenT) [inline soc/renesas/ra/ra2/Kconfig.soc:4]: Should be SOC_SERIES_RA2L1. This is a common issue, please fix all similar cases.

File: soc/renesas/ra/ra2/Kconfig
Comment (thaoluonguw) [inline soc/renesas/ra/ra2/Kconfig:4]: We don't recommend using "ra2" as SoC series because there is risky for future MCUs. We considered for development of future RA MCU and recommend using RA2L1 as SoC Series:
SoC Family: RA
SoC Series: RA8M1, RA8D1, RA8T1, RA6M1, .... RA2L1
SoC: R7FA2L1AB3CFP, R7FA2L1AB3CFP, ...
Board: EK-RA8M1, EK-RA2L1, ....

File: boards/renesas/ek_ra2l1/ek_ra2l1.dts
Comment (thaoluonguw) [inline boards/renesas/ek_ra2l1/ek_ra2l1.dts:131]: May I confirm that why you don't use RA_PSEL?

**Review by thaoluonguw — COMMENTED** [review]
Hello @avolkov-1221, It's really huge PR. If it is possible, please separated it into small for easy review.

File: drivers/serial/uart_renesas_ra2_sci.c
Comment (KhiemNguyenT) [inline drivers/serial/uart_renesas_ra2_sci.c:1]: There is an existing UART driver naming uart_renesas_ra_sci.c , why not using that driver ? If some features need to be added, please clarify.

**Review by KhiemNguyenT — CHANGES_REQUESTED** [review]
@avolkov-1221 , I went through the PR and saw some changes needed. such as naming convention for folder structure, the CONFIG. Additionally, I left the question about reusing existing driver for serial driver. Same question for GPIO, Pinctrl.
Please also consider to add a boards/renesas/ek_ra2l1/doc folder with some description about the board.

IMO, this PR is too big. Please divide it into smaller PRs.

**Comment by avolkov-1221** [comment]
I apologize to everyone (especially @KhiemNguyenT  and @ydamigos), I agree that I crossed the line.  I'll trying this discussion as neutral as it possible, without any personalities.

File: soc/renesas/ra/ra2l1/ofs_mem.c
Comment (soburi) [inline soc/renesas/ra/ra2l1/ofs_mem.c:1]: I don't want to have multiple implementations, can you just do it the way RA4M1 does?
Or, if your implementation is better, can you update RA4M1?

https://github.com/zephyrproject-rtos/zephyr/blob/14b3293d94a1997d0d95dc5b03317c13b3854881/soc/renesas/ra/ra4m1/soc.c#L78

File: drivers/serial/Kconfig.renesas_ra2
Comment (soburi) [inline drivers/serial/Kconfig.renesas_ra2:1]: Consider integrating it into the existing uart_renesas_ra_sci.c.

**Review by soburi — CHANGES_REQUESTED** [review]

File: drivers/serial/Kconfig.renesas_ra2
Comment (avolkov-1221) [reply drivers/serial/Kconfig.renesas_ra2:1]: I'll try to do it. However it would probably be better to make this as a second iteration (as a separate PR).

File: boards/renesas/ek_ra2l1/ek_ra2l1.dts
Comment (avolkov-1221) [reply boards/renesas/ek_ra2l1/ek_ra2l1.dts:131]: Are you asking why I didn't use the RA_PSEL() macro? 

If yes, then my answer is simple: because I don't need to. This pinmux syntax is more readable than the macro-based one (you can immediately understand the purpose of the fields), controlled by yaml, allow to use fields names in the drivers, and is common for the Linux kernel too.

File: soc/renesas/ra/ra2/Kconfig.soc
Comment (avolkov-1221) [reply soc/renesas/ra/ra2/Kconfig.soc]: Completed. In the new patch series.

File: soc/renesas/ra/ra2/Kconfig
Comment (avolkov-1221) [reply soc/renesas/ra/ra2/Kconfig]: Completed. In the latest patch series.

File: soc/renesas/ra/ra2/Kconfig
Comment (avolkov-1221) [reply soc/renesas/ra/ra2/Kconfig]: Done.

File: soc/renesas/ra/ra2/Kconfig
Comment (avolkov-1221) [reply soc/renesas/ra/ra2/Kconfig]: Done.

File: drivers/lpm/Kconfig
Comment (avolkov-1221) [reply drivers/lpm/Kconfig]: Ok. I'll think about how to implement it better. However, I'm not sure yet if the PM API allows this to be done easily (the LPM needs module id as an argument).

File: boards/renesas/ek_ra2l1/board.yml
Comment (avolkov-1221) [reply boards/renesas/ek_ra2l1/board.yml]: Fixed in the latest commit

**Review by soburi — CHANGES_REQUESTED** [review]

File: dts/arm/renesas/ra/ra2/ra2l1.dtsi
Comment (ydamigos) [inline dts/arm/renesas/ra/ra2/ra2l1.dtsi:1]: There are many nodes without the respective DTS bindings, so it is not possible to crosscheck them.
Could you introduce the nodes when you add the drivers in follow-up PRs?

**Review by ydamigos — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_renesas_ra2.c
Comment (ydamigos) [inline drivers/gpio/gpio_renesas_ra2.c:131]: `gpio_port_set_masked_raw()` may also set a pin to low physical level. Does the current implementation support this?

File: drivers/adc/adc_renesas_ra.c
Comment (avolkov-1221) [reply drivers/adc/adc_renesas_ra.c:351]: Sure. Done. The PR is https://github.com/zephyrproject-rtos/zephyr/pull/77555

File: drivers/clock_control/Kconfig
Comment (avolkov-1221) [reply drivers/clock_control/Kconfig:87]: About the clocks:  I'm really thinking about how to better utilize both of them. So far, it's not very straightforward: one of the ideas used in our implementation is the use of fine-grained ref counters, which would allow the PM to switch off unused clocks in sleep modes (the DT support of the counters already in place). A good example of this functionality is when the CAN module is the sole client of the MOSC clock.  However, I don't see such features in the FSP implementations.

**Review by duynguyenxa — CHANGES_REQUESTED** [review]

File: boards/renesas/ek_ra2l1/ek_ra2l1_defconfig
Comment (avolkov-1221) [reply boards/renesas/ek_ra2l1/ek_ra2l1_defconfig:4]: Oops, thank you.

File: soc/renesas/ra/soc.yml
Comment (KhiemNguyenT) [inline soc/renesas/ra/soc.yml:40]: Please move it up, after ra2a1.

**Review by KhiemNguyenT — CHANGES_REQUESTED** [review]
It sounds to me that some improvement feedback has not been addressed, esp. reuse the existing drivers rather than adding the new ones. And there's no explanation in review comments or in commit messages. 
@avolkov-1221 please understand that RA2L1 will not stand alone. RA2A1 is one of its brothers and others will come. Your contribution to existing drivers and next ones are welcomed, but the implementation must be aligned across RA2 devices and other RA series, if there's no big reason to create a new ones.

To accelerate the review and get the patches merged, please consider a " PR for minimal support", without any changes which still need more discussion, dependency to another PR. Next ones should be driver support, such as LPM, ICU, etc. If you want to get the PR merge quickly, please consider to split the PRs.

Last but not least, this PR consumed much effort of many reviewers. I hope we can have focus on smaller PR from now. This is a reasonable request, right ?

File: soc/renesas/ra/soc.yml
Comment (avolkov-1221) [reply soc/renesas/ra/soc.yml:40]: Done. Thank you. 

BTW, could your team update the other SoCs (I mean RA8xx and RA6xx) according to this commit?

In my opinion, using the temperature and quality grade information in the names of the SoCs is useless. However, the socket type (referred to as "package type" in datasheets) defines the number of GPIOs and some peripherals, making it meaningful from a software perspective. Additionally, this can't be easily defined in DTS.

**Comment by avolkov-1221** [comment]
> It sounds to me that some improvement feedback has not been addressed, esp. reuse the existing drivers rather than adding the new ones. And there's no explanation in review comments or in commit messages. @avolkov-1221 please understand that RA2L1 will not stand alone. RA2A1 is one of its brothers and others will come. Your contribution to existing drivers and next ones are welcomed, but the implementation must be aligned across RA2 devices and other RA series, if there's no big reason to create a new ones.

I haven't started the merge yet, except for the OFS, clocks, GPIO, and pinctrls, so I didn't have anything to reply to or comment on.

> 
> To accelerate the review and get the patches merged, please consider a " PR for minimal support", without any changes which still need more discussion, dependency to another PR. Next ones should be driver support, such as LPM, ICU, etc. If you want to get the PR merge quickly, please consider to split the PRs.

Well, it seems reasonable to separate the first commit from the others. But for the rest, that's not the case. They depend on each other.

> 
> Last but not least, this PR consumed much effort of many reviewers. I hope we can have focus on smaller PR from now. This is a reasonable request, right ?

Yes, I can freeze this PR and create new ones. Perhaps they won't be so ill-fated.

File: drivers/clock_control/Kconfig
Comment (soburi) [reply drivers/clock_control/Kconfig:87]: > About the clocks: I'm really thinking about how to better utilize both of them. So far, it's not very straightforward: one of the ideas used in our implementation is the use of fine-grained ref counters, which would allow the PM to switch off unused clocks in sleep modes (the DT support of the counters already in place). A good example of this functionality is when the CAN module is the sole client of the MOSC clock. However, I don't see such features in the FSP implementations.

Understood.
If you still need to consider whether it's optimal for other SoCs, I think it's fine to leave it as it is for now.
However, ultimately, there should be a single implementation for the RA family, so please design it assuming it will be integrated.

**Pull request closed without merging**
